### ISSUE TITLE
Bootstrapping, constructors, and hooks

### ISSUE THREAD

**Issue by justintadlock** [WPTRT/admin-notices#3]
This should perhaps be broken into a couple of different tickets, but I think it all ties together to some degree.  It's also some thoughts on the bigger picture as we develop more packages in the future.

Generally speaking, the `__construct()` method should not call `add_action()` as done here: https://github.com/WPTRT/admin-notices/blob/master/src/Notices.php#L40-L47.  The constructor's only job should be setting up object properties if needed.

Instead, action/filter hook calls should happen in a separate method that I'd name `boot()` like so:

```php
public function boot() {
	// add_action() and add_filter() calls go here.
}
```

We'll dig into the "why" of this, which I hope will become clearer by the end.  But, first, let's look at the bootstrapping code in its current state:

```php
$notices = new \WPTRT\AdminNotices\Notices();

// Add a notice.
$notices->add( (string) $id, (string) $title, (string) $content, (array) $options );
```

I do worry that theme authors might just drop this anywhere in `functions.php` and have a global `$notices` variable with that (we might need to caution against that).  But, it mostly looks good.

The issue for me is different.  In a real-world project, I'd ideally register the `Notices` class as a single instance with a container like so:

```php
// Register early.
$container->singleton( Notices::class );

// Bootstrap later when needed.
$container->resolve( Notices::class )->boot();
```

The problem here is that by using `__construct()` to run `add_action()` it's combining setting up the object and executing actions.  There's no way I can wait to bootstrap the actions.  And, child themes cannot simply remove the `Notices` instance from the container between the point of registration and bootstrapping.

I think bootstrapping would work better as:

```php
// Create a new Notices instance.
$notices = new \WPTRT\AdminNotices\Notices();

// Add a notice.
$notices->add( (string) $id, (string) $title, (string) $content, (array) $options );

// Bootstrap notice actions.
$notices->boot();
```

---

On a "larger picture" note that goes beyond this ticket and beyond this particular TRT package, I'm not sure how I feel about combining both the "collection" of notices with the actions/filters.  I've struggled with how to best simplify that and whether there should be an additional "component" or "manager" class.  I feel like that might overcomplicate things for something that we're trying to keep simple for theme authors.

Basically, it comes down to separation of concerns.  `Notices` is doing two jobs.  _Should it be concerned with anything other than storing `Notice` objects?_  And, this may not be a big concern for this project specifically.  Just posing the question.

Anyway, something to think about.

**Comment by aristath** [comment]
@justintadlock While I agree with the above, I think for the purposes of this package it's important to keep things simple for developers.
I moved the actions outside of the constructor, but again the method created to contain them runs directly in the constructor.

However, in a _real/big_ project like you say I'd probably follow a slightly different approach too.
A simple solution would be to simply bypass the `Notices` object using something like this:
```php
$my_notices = apply_filters( 'my_notices', [] );
add_action( 'admin_notices', function() use ( $my_notices ) {
	foreach ( $my_notices as $notice ) {
		$notice->the_notice();
	}
});
add_action( 'admin_footer', function() use ( $my_notices ) {
	foreach ( $my_notices as $notice ) {
		$notice->dismiss->print_script();
	}
});
```
The above completely replaces the `Notices` object and we can then use `Notice` directly using the filter from above:
```php
add_filter( 'my_notices', function( $notices ) {
	$notices[] = new Notice( 'just_a_notice', 'My Title', 'The Message' );
	return $notices;
} );
```
Of course that's just some simple sample code, but it introduces a `my_notices` filter that can then be used to handle everything.

The `Notice` object creates both the content and the script, so in case something more complicated is required the `Notices` class can just be skipped. It's just a convenient way to make things easier to manage for most use cases. I doubt something like that will ever be needed in a theme on .org, but even if it's required it's pretty simple to do using the `Notice` objects which are pretty-much self-contained.

Simple use-cases can use the conveniency of `Notices`, if someone wants a more complicated solution they'll just need to write a few more lines of code but it's still relatively easy 👍

**Comment by justintadlock** [comment]
Calling `$this->run()` in `__construct()` is really no different from before.

Constructors should be used for setting object properties. I can't think of a good reason to automatically attach actions to hooks other than saving a single line of code for the theme author.  We're just constructing a new object at that point.  There's no need to perform any other action.

I'd remove the `__construct()` call in the `Notices` class.  The developer should call `$notices->run()` after adding their notices.

The one extra line of code doesn't take away from the simplicity.  It also wouldn't require any extra custom coding on the part of someone using something like a container or doing something different.

**Comment by aristath** [comment]
> Calling $this->run() in __construct() is really no different from before.

I know... it just looks better.

> The one extra line of code doesn't take away from the simplicity. It also wouldn't require any extra custom coding on the part of someone using something like a container or doing something different.

Fair point. I'll push that in a few minutes 👍

**Comment by aristath** [comment]
@justintadlock I forgot to mention I pushed this change yesterday and tweaked the readme file accordingly.

I think we should be OK to tag v1.0 if you agree 👍

**Comment by justintadlock** [comment]
Looks good to me!
